keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -228,6 +225,22 @@ kind: CustomResourceDefinition
               type: integer
 ```
 
+Example Validation Rules:
+
+| Rule                                                                             | Purpose                                                                           |
+| ----------------                                                                 | ------------                                                                      |
+| `self.minReplicas <= self.replicas <= self.maxReplicas`                          | Validate that the three fields defining replicas are ordered appropriately        |
+| `'Available' in self.stateCounts`                                                | Validate that an entry with the 'Available' key exists in a map                   |
+| `(size(self.list1) == 0) != (size(self.list2) == 0)`                             | Validate that one of two lists is non-empty, but not both                         |
+| `has(self.expired) && self.created + self.ttl < self.expired`                    | Validate that 'expired' date is after a 'create' date plus a 'ttl' duration       |
+| `self.health.startsWith('ok')`                                                   | Validate a 'health' string field has the prefix 'ok'                              |
+| `self.widgets.exists(w, w.key == 'x' && w.foo < 10)`                             | Validate that the 'foo' property of a listMap item with a key 'x' is less than 10 |
+| `type(self.limit) == string ? self.limit == '100%' : self.limit == 1000`         | Validate an int-or-string field for both the the int and string cases             |
```

Comment:
Since this isn't doing any cross-field validation, it could also be placed on the int-or-string itself and written like this, right?

```suggestion
| `type(self) == string ? self == '100%' : self == 1000`         | Validate an int-or-string field for both the the int and string cases             |
```

It might be worth including a bare `self` example in this list

Reply:
I also didn't see it explicitly stated whether validation expressions defined on a property are evaluated if that property is not present in the data being validated.

For example, if I define an optional integer property `foo` with a validation expression `self > 10`, then submit an object lacking the property `foo` entirely, I would expect my validation expression to not run, and the object to be accepted.

If I wanted to require `foo` to be present and to be greater than 10, I have a few options:
* mark `foo` (and all ancestor properties) as required, and absence would complain about a missing required field (but still wouldn't evaluate my validation expression)... once I added `foo` to my object, my expression defined on `foo` would be evaluated
* put my validation expression on a required parent of `foo`, e.g. put it at the top level as `self.spec.foo > 10`, and my expression would be evaluated even if foo was not present in the object.

Is that understanding correct?

Reply:
Yes, you are correct for both cases. I'll add this to the KEP.